README
NAME
    Parse::DMIDecode - Interface to SMBIOS under Linux using dmidecode

SYNOPSIS
     use strict;
     use Parse::DMIDecode ();
     
 my $decoder = new Parse::DMIDecode;
     $decoder->probe; # Actively probe using dmidecode
      
 # Manually supply your own dmidecode output to be parsed
     # $decoder->parse(qx(sudo /usr/sbin/dmidecode));
      
 printf("System: %s, %s",
             $decoder->keyword("system-manufacturer"),
             $decoder->keyword("system-product-name"),
         );

DESCRIPTION
    This module provides an OO interface to SMBIOS information through the
    *dmidecode* command which is known to work under a number of Linux, BSD
    and BeOS variants.

    This module is still actively under development, is no not yet feature
    complete and still needs to be fully documented. There is a possibility
    accessor method names may change before the final release.

METHODS
  new
     my $decoder = Parse::DMIDecode->new(
                         dmidecode => "/usr/sbin/dmidecode"),
                         nowarnings => 1
                     );

  probe
     $decoder->probe;

  parse
     my $raw = qx(sudo /usr/sbin/dmidecode);
     $decoder->prase($raw);

  keyword
     my $serial_number = $decoder->keyword("system-serial-number");

  keywords
     my @keywords = $decoder->keywords;
     my @bios_keywords = $decoder->keywords("bios");
     
 for my $keyword (@bios_keywords) {
         printf("%s => %s\n",
                 $keyword,
                 $decoder->keyword($keyword)
             );
     }

  handle_addresses
     my @addresses = $decoder->handle_addresses;

  get_handles
     use Parse::DMIDecode::Constants qw(@TYPES);
     
 # Available groups to query: bios, system, baseboard,
     #    chassis, processor, memory, cache, connector, slot
     for my $handle ($decoder->get_handles( group => "memory" )) {
         printf(">> Found handle at %s (%s):\n%s\n",
                 $handle->address,
                 $TYPES[$handle->dmitype],
                 $handle->raw
             );
     }

    See Parse::DMIDecode::Handle for accessor method documentation for
    handle objects.

  smbios_version
     my $smbios_version = $decoder->smbios_version;

  dmidecode_version
     my $dmidecode_version = $decoder->dmidecode_version;

    Returns the version number of the copy of *dmidecode* that was used to
    create the source data that was parsed. This value may not be available
    when using older versions of *dmidecode*.

  table_location
     my $memory_address = $decoder->table_location;

  structures
     my $total_structures = $decoder->structures;

SEE ALSO
    Parse::DMIDecode::Handle, Parse::DMIDecode::Constants, examples/*.pl,
    <http://www.nongnu.org/dmidecode/>, <http://linux.dell.com/libsmbios/>,
    <http://sourceforge.net/projects/x86info/>,
    <http://www.dmtf.org/standards/smbios>, biosdecode(8), dmidecode(8),
    vpddecode(8)

VERSION
    $Id$

AUTHOR
    Nicola Worthington <nicolaw@cpan.org>

    <http://perlgirl.org.uk>

    If you like this software, why not show your appreciation by sending the
    author something nice from her Amazon wishlist? (
    http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

COPYRIGHT
    Copyright 2006 Nicola Worthington.

    This software is licensed under The Apache Software License, Version
    2.0.

    <http://www.apache.org/licenses/LICENSE-2.0>



Changes
Version 0.04 - 19th March 2007
    + Fixed a few typos in the POD.


Version 0.03 - 11th March 2007
    + Added Parse::DMIDecode::Examples POD file containing
      a small selection of useful code examples.

    + Changed Build.PL to be more paranoid defaulting
      to no while using $build->y_n to work around
      potential removal of expected method functionality.

    + Added complain.txt.

    + Removed depencancy on Scalar::Util.

    + Added some additional POD.


Version 0.02 - 22th October 2006
    + Added better support for older versions of dmidecode
      binary output parsing.

    + Improved examples/test.pl.

    + Added examples/synopsis2.pl from the synopsis of
      Parse::DMIDecode::Handle.

    + Allow get_handles() method to be called without any
      qualifying paramaters so to return *All* handle
      objects.

    + Added description() method to handle objects.

    + Added parsed_structures() method to handle objects.

    + Improved parsing inside handle objects to return
      multiple handle objects where appropriate to allow
      easier access to unique data pairs.


Version 0.01 - 19th October 2006
    + Pre-release.

    + Created Parse::DMIDecode::Handle objects for each
      handle structure.

    + Encapsulated the internal data structure which is
      quite fragile so that people cannot just grab items
      out of it.

    + Added accessor methods for header information.

    + Added selection methods to return handle objects.

    + Added a method to return a list of all known
      keyword strings.

    + Added unit tests.

    + Added a whole bunch more since the first release
      was mostly a placeholder.


Version 0.00 - 17th September 2006
    + Initial developer pre-release to save the namespace
      and to discuss API prior to full release.



examples/test.pl
#!/usr/bin/perl -w

#BEGIN { chdir '../' if -d '../examples/'; }

use strict;
use lib qw(./lib ../lib);
use Getopt::Std qw(getopts);
use Fcntl qw(:mode);
use Parse::DMIDecode;
use Parse::DMIDecode::Constants qw(@TYPES);

# Get some command line options
my $opts = {};
Getopt::Std::getopts('f:g:Kk:hV',$opts);
$opts->{g} ||= 'processor';
display_help(),exit if defined $opts->{h};

my $decoder = new Parse::DMIDecode;
my $dmidecode = '/usr/sbin/dmidecode';
my @stat = stat($dmidecode);

# If we've been given a dmidecode output file, parse that
if ($opts->{f}) {
	die "File '$opts->{f}' does not exist.\n" unless -f $opts->{f};
	$decoder->parse(`cat $opts->{f}`);

# If we're root, or dmidecode is setuid root, then probe
} elsif ($> || ($stat[4] == 0 && $stat[2] & S_ISUID)) {
	$decoder->probe;

# Otherwise run dmidecode with sudo
} else {
	$decoder->parse(qx(sudo $dmidecode));
}

# Print all of the available keywords
if (defined $opts->{K}) {
	if (defined $opts->{V}) {
		for my $keyword ($decoder->keywords) {
			my $value = $decoder->keyword($keyword);
			$value = '' unless defined $value;
			printf("Keyword '%s' => '%s'\n",
					$keyword,
					(ref($value) eq 'ARRAY' ? join(', ',@{$value}) : $value)
				);
		}
	} else {
		print join("\n",$decoder->keywords)."\n";
	}
	exit;

# Print just one keyword
} elsif ($opts->{k}) {
	printf("Keyword '%s' => '%s'\n",
			$opts->{k},
			$decoder->keyword($opts->{k})
		);
	exit;
}

# Print some information about specific structure handles
for my $handle ($decoder->get_handles( group => $opts->{g} )) {
	printf(">>> Found handle at %s (%s):\n >> Keywords: %s\n%s\n",
			$handle->address,
			$TYPES[$handle->dmitype],
			join(', ',$handle->keywords),
			$handle->raw
		);
	if (defined $opts->{V}) {
		for my $keyword ($handle->keywords) {
			my $value = $handle->keyword($keyword);
			$value = '' unless defined $value;
			printf("  > Keyword '%s' => '%s'\n",
					$keyword,
					(ref($value) eq 'ARRAY' ? join(', ',@{$value}) : $value)
				);
		}
	}
	print "\n";
}

sub display_help {
	print qq{Syntax: $0 [-h] [-V] [-K|-k <keyword>|-g <group>] [-f <filename>]
    -h              Display this help
    -K              List all valid keywords
    -V              Print more verbose output
    -g <group>      Display handle information match <group> group
    -f <filename>   Parse dmidecode information from <filename>
};
}




Build.PL
# vim:ts=4:sw=4:tw=78
# $Id$

use strict;
use Module::Build;
use vars qw($build);

my $module = 'Parse::DMIDecode';
$build = Module::Build->new(
		module_name        => $module,
		license            => 'open_source',
		create_makefile_pl => 'passthrough',
		create_readme      => 1,
		create_packlist    => 1,
		sign               => 0,

		requires => {
			'Carp'         => 0, # 1.04
			#'Scalar::Util' => 0,
			#'Storable'     => 0,
		},

		build_requires => {
			'Test'         => 0, # 1.25
			'Test::More'   => 0, # 0.60
		},

		recommends => {
			'File::Which'         => 0.05, # 0.05
			'Test::Pod'           => 1.20,
			'Test::Pod::Coverage' => 1.06,
			'Test::Deep'          => 0.093,
		},
	);

$build->create_build_script;


# Send perl and module version information home if we've been given
# permission to do so by a human being - default to not send for automated
# testing environments, of if the user does not respond within 20 seconds.

my $url = $ENV{AUTOMATED_TESTING} ? undef : may_send_version_information();
if ($url) {
	my @resp = ();
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 10;
		my $ua = LWP::UserAgent->new(
				agent => 'Build.PL $Revision$',
				timeout => 9,
				max_size => 500,
			);
		$ua->env_proxy;
		my $response = $ua->get($url);
		if ($response->is_success()) {
			for (split(/\s*\n+\s*/, $response->content())) {
				push @resp, $_ if $_;
			}
		}
		alarm 0;
	};
	print substr($resp[0],0,79) || "Thank you for sending this information.";
	print "\n\n";
}

sub may_send_version_information {
	eval {
		require Config;
		require LWP::UserAgent;
	};
	return undef if $@;

	my $str = sprintf('%s?%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s',
			'http://perlgirl.org.uk/lib/usage.cgi',
			'name',     $module,
			'version',  $build->dist_version(),
			'osname',   $Config::Config{osname},
			'archname', $Config::Config{archname},
			'osver',    $^O,
			'perlver',  $]
		);

	print "\nThank you for downloading ".$build->dist_name()."\n\n";
	print "I would like to find out how many people are using this software,\n";
	print "and on what operating systems and Perl versions. If you have an\n";
	print "internet connection, may I transmit the following information:\n\n";
	print "$str\n\n";

	my $send = 0;
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 20;
		$send = $build->y_n('Send this anonymous information?','n');
		alarm 0;
	};

	return defined $send && !ref($send) && "$send" eq "1" ? $str : undef;
}

1;



DistBuild.PL
#!/usr/bin/env perl
############################################################
#
#   $Id$
#   DistBuild.PL - CPAN Distribution Build Script for Subversion & CVS
#
#   Copyright 2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################
# vim:ts=4:sw=4:tw=78

use 5.6.1;
use strict;
use warnings;
use vars qw($build);

$build = undef;
do 'Build.PL';

check_repository();

if ($ENV{CVSROOT} && -d 'CVS' && !-d '.svn' && `which cvs2cl.pl`) {
	stage('Updating ChangeLog');
	$build->do_system('cvs2cl.pl', qw(-r -t -T -P --fsf --no-times))
		&& unlink 'ChangeLog.bak';
}

unlink 'MANIFEST';
stage('Updating MANIFEST');
$build->dispatch('manifest');

stage('Updating META.yml');
$build->dispatch('distmeta');

stage('Building');
$build->dispatch('build');

stage('Checking for essential files');
check_essential_files();

stage('Testing');
$build->dispatch('test', verbose => 0);

stage('Installing');
$build->dispatch('install');

stage('Building distribution tarball');
my @tarballs = ();
if (check_repository(uncommited_only => 1, silent => 1, ignore => "(META.yml|Makefile.PL|README|blib|_build|Build|MANIFEST|tarballs/)")) {
	print "Uncommited files; skipping ...\n";
} else {
	$build->dispatch('distcheck');
	chmod 0644, 'Build.PL';
	$build->dispatch('dist');
	$build->dispatch('ppmdist');

	stage('Moving distribution tarball');
	@tarballs = move_dist_tarballs();
}

stage('Cleaning up');
$build->dispatch('distclean');

check_repository(uncommited_only => 1);

print "\n";
exit;



sub check_repository {
	my %opts = @_;
	my $uncommited_files = 0;
	$opts{ignore} ||= '(META.yml|Makefile.PL|README|blib|_build|Build|MANIFEST)';

	my ($cvs,$svn) = (0,0);
	if (-d 'CVS' && !-d '.svn') {
		$cvs = 1;
	} elsif (-d '.svn' && !-d 'CVS') {
		$svn = 1;
	} elsif (-d '.svn' || -d 'CVS') {
		die "Found both ./.svn/ and ./CVS/; way too confusing!\n";
	} else {
		warn "Couldn't find ./.svn/ or ./CVS/; skipping repository checks ...\n"
	}

	if ($cvs) {
		stage('Checking file statuses in CVS ...') unless $opts{no_stage};
		my $cmd = sprintf("cvs -q status | grep File: | %s egrep -v '%s' %s",
				($opts{uncommited_only} ? ' grep -v "Status: Up-to-date" | ' : ''),
				$opts{ignore},
				($opts{silent} ? ' >/dev/null 2>&1 ' : ''),
			);
		my $rtn = $build->do_system($cmd);
		$uncommited_files = $rtn if $opts{uncommited_only};

	} elsif ($svn) {
		stage('Checking file statuses in Subversion ...') unless $opts{no_stage};
		my $cmd = sprintf("svn status | egrep -v '%s' %s",
				$opts{ignore},
				($opts{silent} ? ' >/dev/null 2>&1 ' : ''),
			);
		my $rtn = $build->do_system($cmd);
		$uncommited_files = $rtn;
	}

	return $uncommited_files;
}


sub stage {
	local $_ = shift;
	chomp;
	print "\n\n".('*'x60).
		"\n* $_\n".
		('*'x60)."\n";
}


sub check_essential_files {
	my @essential_files = qw(META.yml Build.PL Makefile.PL
			LICENSE NOTICE README Changes INSTALL MANIFEST TODO
			lib t examples);

	my @missing_files = ();
	for (@essential_files) {
		push @missing_files, $_ if !-e $_;
	}

	if (@missing_files) {
		die "Missing: ".join(', ',@missing_files)."\n";
	} else {
		print "No problems found\n";
	}
}


sub move_dist_tarballs {
	my @file = glob('*.tar.gz');
	push @file, glob('*.ppd');
	my @tarballs = ();
	for (@file) {
		print "Moving $_ to tarballs/$_\n";
		#unlink "tarballs/$_" if -e "tarballs/$_";
		rename $_, "tarballs/$_";
		push @tarballs, "tarballs/$_" if -f "tarballs/$_";
	}
	return @tarballs;
}


BEGIN {
	$ENV{AUTOMATED_TESTING} = 1;
	for (qw(MANIFEST.bak),glob('*.tar.gz'),glob('*.tar'),glob('*.ppd')) {
		unlink($_) if defined($_) && -e $_;
	}
	system('find . -maxdepth 3 -type f -name "*.rrd" | xargs rm -fv');
}


END {
	for (qw(MANIFEST MANIFEST.bak),glob('*.tar.gz'),glob('*.tar'),glob('*.ppd')) {
		unlink($_) if defined($_) && -e $_;
	}
}


__END__





lib/Parse/DMIDecode/Constants.pm
############################################################
#
#   $Id$
#   Parse::DMIDecode::Constants - SMBIOS Constants
#
#   Copyright 2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Parse::DMIDecode::Constants;
# vim:ts=4:sw=4:tw=78

use strict;
require Exporter;
use vars qw($VERSION $DEBUG
		@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS
		@TYPES %GROUPS %TYPE2GROUP);

@ISA = qw(Exporter);
@EXPORT = qw();
@EXPORT_OK = qw(@TYPES %GROUPS %TYPE2GROUP);
%EXPORT_TAGS = (all => \@EXPORT_OK);

$VERSION = '0.03' || sprintf('%d', q$Revision$ =~ /(\d+)/g);
$DEBUG ||= $ENV{DEBUG} ? 1 : 0;

@TYPES = ( # Description                    Index Group(s)
		'BIOS',                             # 0   bios
		'System',                           # 1   system
		'Base Board',                       # 2   baseboard
		'Chassis',                          # 3   chassis
		'Processor',                        # 4   processor
		'Memory Controller',                # 5   memory
		'Memory Module',                    # 6   memory
		'Cache',                            # 7   cache
		'Port Connector',                   # 8   connector
		'System Slots',                     # 9   slot
		'On Board Devices',                 # 10  baseboard
		'OEM Strings',                      # 11
		'System Configuration Options',     # 12  system
		'BIOS Language',                    # 13  bios
		'Group Associations',               # 14
		'System Event Log',                 # 15  system
		'Physical Memory Array',            # 16  memory
		'Memory Device',                    # 17  memory
		'32-bit Memory Error',              # 18
		'Memory Array Mapped Address',      # 19
		'Memory Device Mapped Address',     # 20
		'Built-in Pointing Device',         # 21
		'Portable Battery',                 # 22
		'System Reset',                     # 23  system
		'Hardware Security',                # 24
		'System Power Controls',            # 25
		'Voltage Probe',                    # 26
		'Cooling Device',                   # 27
		'Temperature Probe',                # 28
		'Electrical Current Probe',         # 29
		'Out-of-band Remote Access',        # 30
		'Boot Integrity Services',          # 31
		'System Boot',                      # 32  system
		'64-bit Memory Error',              # 33
		'Management Device',                # 34
		'Management Device Component',      # 35
		'Management Device Threshold Data', # 36 
		'Memory Channel',                   # 37
		'IPMI Device',                      # 38
		'Power Supply'                      # 39
	);

%GROUPS = (
		'bios'      => [ qw(0 13) ],
		'system'    => [ qw(1 12 15 23 32) ],
		'baseboard' => [ qw(2 10) ],
		'chassis'   => [ qw(3) ],
		'processor' => [ qw(4) ],
		'memory'    => [ qw(5 6 16 17) ],
		'cache'     => [ qw(7) ],
		'connector' => [ qw(8) ],
		'slot'      => [ qw(9) ],
	);

%TYPE2GROUP = ();
for my $group (keys %GROUPS) {
	for my $dmitype (@{$GROUPS{$group}}) {
		$TYPE2GROUP{$dmitype} = $group;
	}
}


1;


=pod

=head1 NAME

Parse::DMIDecode::Constants - SMBIOS Constants

=head1 SYNOPSIS

 use strict;
 use Parse::DMIDecode::Constants qw(@TYPES %GROUPS);
 
=head1 DESCRIPTION

This module provides and can export constants relating to the SMBIOS
specification and dmidecode interface command.

=head1 EXPORTS

=head2 @TYPES

=head2 %GROUPS

=head1 SEE ALSO

L<Parse::DMIDecode>

=head1 VERSION

$Id$

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

If you like this software, why not show your appreciation by sending the
author something nice from her
L<Amazon wishlist|http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority>? 
( http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

=head1 COPYRIGHT

Copyright 2006 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut


__END__





lib/Linux/SMBIOS.pm
############################################################
#
#   $Id$
#   Linux::SMBIOS - Interface to SMBIOS under Linux using dmidecode
#
#   Copyright 2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Linux::SMBIOS;
# vim:ts=4:sw=4:tw=78

use strict;
use Carp qw(croak cluck confess carp);
use vars qw($VERSION $DEBUG);

$VERSION = '0.01' || sprintf('%d', q$Revision$ =~ /(\d+)/g);
$DEBUG ||= $ENV{DEBUG} ? 1 : 0;


sub new {
	ref(my $class = shift) && croak 'Class name required';
	croak 'Odd number of elements passed when even was expected' if @_ % 2;
	my $self = { @_ };

	my @commands = qw(dmidecode biosdecode x86info);
	my $validkeys = join('|',@commands);
	cluck('Unrecognised parameters passed: '.
		join(', ',grep(!/^$validkeys$/,keys %{$self})))
		if (grep(!/^$validkeys$/,keys %{$self}) && $^W);

	for my $command (@commands) {
		croak "Command $command '$self->{$command}'; file not found"
			if defined $self->{$command} && !-f $self->{$command};
	}

	if (!defined $self->{dmidecode}) {
		require File::Which;
		for my $command (@commands) {
			$self->{$command} = File::Which::which($command)
				if !defined $self->{$command};
		}
	}

	bless($self,$class);
	DUMP($class,$self);
	return $self;
}


sub probe {
	my $self = shift;
	unless (ref $self && UNIVERSAL::isa($self, __PACKAGE__)) {
		unshift @_, $self unless $self eq __PACKAGE__;
		$self = new __PACKAGE__;
	}

	my ($cmd) = $self->{dmidecode} =~ /^([\/\.\_\-a-zA-Z0-9 >]+)$/;
	TRACE($cmd);

	my $fh;
	open($fh,'-|',$cmd) || croak "Unable to open file handle for command '$cmd': $!";
	while (local $_ = <$fh>) {
		$self->{dmidecode_data} .= $_;
	}
	close($fh) || carp "Unable to close file handle for command '$cmd': $!";
	$self->{dmidecode} = $self->_parse('dmidecode',$self->{dmidecode_data});

	DUMP('$self',$self);
	return scalar($self->{dmidecode}) ?
			scalar(keys(%{$self->{dmidecode}->{strct}})) : undef;
}


sub _parse {
	my ($self,$type,$str) = @_;

	my %strct;
	my %data;
	my $handle = '';
	my $key = '';
	my $indent = '';

	for (split(/\n/,$str)) {
		next if /^\s*$/;

		if (/^# dmidecode ([\d\.]+)\s*$/) {
			$data{dmidecode} = $1;

		} elsif (/^(\d+) structures occupying (\d+) bytes?\.\s*$/) {
			$data{structures} = $1;
			$data{bytes} = $2;

		} elsif (/^SMBIOS ([\d\.]+) present\.?\s*$/) {
			$data{smbios} = $1;

		} elsif (/^Table at (0-9A-Fx]+)\.?\s*$/) {
			$data{location} = $1;

		} elsif (/^Handle ([0-9A-Fx]+)(?:, DMI type (\d+), (\d+) bytes?\.?)?\s*$/) {
			$handle = $1;
			$key = '';
			$strct{$handle} = {};
			$strct{$handle}->{dmitype} = $2 if defined $2;
			$strct{$handle}->{bytes} = $3 if defined $3;

		} elsif (defined $handle && $handle =~ /\S+/) {
			if (/^\s*DMI type (\d+), (\d+) bytes?\.?\s*$/) {
				$strct{$handle}->{dmitype} = $1 if defined $1;
				$strct{$handle}->{bytes} = $2 if defined $2;

			} elsif (defined $strct{$handle}->{dmitype} &&
					!defined $strct{$handle}->{name} &&
					/^\s*(\S+.*?)\s*$/) {
				$strct{$handle}->{name} = $1;
				$key = '';

			} elsif (defined $strct{$handle}->{name}) {
				if (/^\s*(.+?): (\S+.*?)\s*$/) {
					$key = '';
					$strct{$handle}->{data}->{$1} = $2;

				} elsif ((!$key || !/^$indent/) && /^\s+(.+?):\s*$/) {
					$key = $1;

				} elsif ($key && /^(\s+)(\S+.*?)\s*$/) {
					$indent = $1;
					my $data = $2;

					if (  (!defined($strct{$handle}->{data}->{$key}) ||
							ref($strct{$handle}->{data}->{$key}) ne 'ARRAY') 
							&& $data =~ /^(?:[0-9A-F]{2} )+(?:[0-9A-F]{2})?$/) {
						$strct{$handle}->{data}->{$key} .=
							defined $strct{$handle}->{data}->{$key} ?
							" $data" : $data;

					} else {
						$strct{$handle}->{data}->{$key} = [()]
							unless defined $strct{$handle}->{data}->{$key};
						push @{$strct{$handle}->{data}->{$key}}, $data;
					}
				}
			}
		}
	}

	$data{strct} = \%strct;
	carp sprintf("Only parsed %d structures when %d were expected",
			scalar(keys(%strct)), $data{structures})
		if scalar(keys(%strct)) != $data{structures};

	DUMP($type,\%data);
	return \%data;
}


sub TRACE {
	return unless $DEBUG;
	carp(shift());
}


sub DUMP {
	return unless $DEBUG;
	eval {
		require Data::Dumper;
		$Data::Dumper::Indent = 2;
		$Data::Dumper::Terse = 1;
		carp(shift().': '.Data::Dumper::Dumper(shift()));
	}
}

1;


=pod

=head1 NAME

Linux::SMBIOS - Interface to SMBIOS under Linux using dmidecode

=head1 SYNOPSIS

 use strict;
 use Linux::SMBIOS ();
 
 # Create an interface object
 my $smbios = new Linux::SMBIOS;

=head1 DESCRIPTION

=head1 METHODS

=head2 new

=head2 probe

=head1 SEE ALSO

examples/*.pl,
L<http://www.nongnu.org/dmidecode/>,
L<http://linux.dell.com/libsmbios/>,
L<http://sourceforge.net/projects/x86info/>,
L<http://www.dmtf.org/standards/smbios>,
L<biosdecode(8)>, L<dmidecode(8)>, L<vpddecode(8)>

=head1 VERSION

$Id$

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

If you like this software, why not show your appreciation by sending the
author something nice from her
L<Amazon wishlist|http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority>? 
( http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

=head1 COPYRIGHT

Copyright 2006 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut


__DATA__

static const u8 opt_type_bios[]={ 0, 13, 255 };
static const u8 opt_type_system[]={ 1, 12, 15, 23, 32, 255 };
static const u8 opt_type_baseboard[]={ 2, 10, 255 };
static const u8 opt_type_chassis[]={ 3, 255 };
static const u8 opt_type_processor[]={ 4, 255 };
static const u8 opt_type_memory[]={ 5, 6, 16, 17, 255 };
static const u8 opt_type_cache[]={ 7, 255 };
static const u8 opt_type_connector[]={ 8, 255 };
static const u8 opt_type_slot[]={ 9, 255 };

static const struct type_keyword opt_type_keyword[]={
	{ "bios", opt_type_bios },
	{ "system", opt_type_system },
	{ "baseboard", opt_type_baseboard },
	{ "chassis", opt_type_chassis },
	{ "processor", opt_type_processor },
	{ "memory", opt_type_memory },
	{ "cache", opt_type_cache },
	{ "connector", opt_type_connector },
	{ "slot", opt_type_slot },
};

static const struct string_keyword opt_string_keyword[]={
	{ "bios-vendor", 0, 0x04, NULL, NULL },
	{ "bios-version", 0, 0x05, NULL, NULL },
	{ "bios-release-date", 0, 0x08, NULL, NULL },
	{ "system-manufacturer", 1, 0x04, NULL, NULL },
	{ "system-product-name", 1, 0x05, NULL, NULL },
	{ "system-version", 1, 0x06, NULL, NULL },
	{ "system-serial-number", 1, 0x07, NULL, NULL },
	{ "system-uuid", 1, 0x08, NULL, dmi_system_uuid },
	{ "baseboard-manufacturer", 2, 0x04, NULL, NULL },
	{ "baseboard-product-name", 2, 0x05, NULL, NULL },
	{ "baseboard-version", 2, 0x06, NULL, NULL },
	{ "baseboard-serial-number", 2, 0x07, NULL, NULL },
	{ "baseboard-asset-tag", 2, 0x08, NULL, NULL },
	{ "chassis-manufacturer", 3, 0x04, NULL, NULL },
	{ "chassis-type", 3, 0x05, dmi_chassis_type, NULL },
	{ "chassis-version", 3, 0x06, NULL, NULL },
	{ "chassis-serial-number", 3, 0x07, NULL, NULL },
	{ "chassis-asset-tag", 3, 0x08, NULL, NULL },
	{ "processor-family", 4, 0x06, dmi_processor_family, NULL },
	{ "processor-manufacturer", 4, 0x07, NULL, NULL },
	{ "processor-version", 4, 0x10, NULL, NULL },
	{ "processor-frequency", 4, 0x16, NULL, dmi_processor_frequency },
};

__END__



